src/llms/llm_loader.py

```diff
@@ -0,0 +1,252 @@
+# workaround to disable UserWarning
+import warnings
+warnings.simplefilter("ignore", UserWarning)
+
+from langchain.callbacks.streaming_stdout import StreamingStdOutCallbackHandler
+from langchain.callbacks.manager import CallbackManager
+
+import os, inspect
+
+## append path of current caller when __main__
+if __name__ == '__main__':
+    import sys 
+    sys.path.insert(0,'')
+
+from utils.logger import Logger
+
+class LLMLoader:
+    """
+    Note: This class loads the LLM backend librearies if the specific LLM is loaded.
+    Known caveats: Currently supports a single instance/model per backend
+    """
+    def __init__(self, llm_backends=set(),
+                 inference_url=None,
+                 prompt_type=None,
+                 api_key=None,
+                 model=None, logger=None) -> None:
+        self.logger = logger if logger is not None else Logger("llm_loader").logger
+        # a set of backend LLMs to activate
+        self.llm_backends = set(os.environ.get('LLM_BACKENDS', {'ollama'})) if len(llm_backends) == 0 else llm_backends
+        # defalt LLM backend to use or use the first index from loaded ones
+        self.llm_default = os.environ.get('LLM_DEFAULT', list(self.llm_backends)[0])
+        self.llm = {}
+        self._set_llm_instance()
+
+    def _set_llm_instance(self, inference_url=None, api_key=None, model=None):
+        self.logger.debug(f"[{inspect.stack()[0][3]}] Loading LLM instances with default {self.llm_default}")
+
+        for backend in self.llm_backends:
+            self.logger.debug(f"Loading backend LLM {backend}")
+            match backend:
+                case 'openai':
+                    self._openai_llm_instance()
+                case 'ollama':
+                    self._ollama_llm_instance()
+                case 'tgi':
+                    self._tgi_llm_instance()
+                case 'watson':
+                    self._watson_llm_instance()
+                case 'bam':
+                    self._bam_llm_instance()
+                case _:
+                    self.logger.warning(f"WARNING: Unsupported LLM {backend}")
+
+    def _openai_llm_instance(self):
+        self.logger.debug(
+            f"[{inspect.stack()[0][3]}] Creating OpenAI LLM instance")
+        try:
+            import openai
+            from langchain.llms import OpenAI 
+        except e:
+            self.logger.error(f"ERROR: Missing openai libraries. Skipping loading backend LLM.")
+            return
+        params = {
+            'base_url': os.environ.get('OPENAI_API_URL', 'https://api.openai.com/v1'),
+            'api_key': os.environ.get('OPENAI_API_KEY', None),
+            'model': os.environ.get('OPENAI_MODEL', None),
+            'model_kwargs': {}, # TODO: add model args
+            'organization': os.environ.get('OPENAI_ORGANIZATION', None),
+            'timeout': os.environ.get('OPENAI_TIMEOUT', None),
```

Comment:
See if the update provides a better alignment

Reply:
it's better in that now a caller can get an instance of a particular LLM.  I still think the configuration/parameterization leaves something to be desired, but since i'm in the middle of working on the next iteration of how we configure this thing, we can leave it be for now.